Lib/typing.py

```diff
@@ -1997,14 +1994,44 @@ def __instancecheck__(cls, instance):
                 val = getattr_static(instance, attr)
             except AttributeError:
                 break
-            # this attribute is set by @runtime_checkable:
             if val is None and attr not in cls.__non_callable_proto_members__:
                 break
         else:
             return True
 
         return False
 
+    @property
+    def __protocol_attrs__(cls):
+        try:
+            return cls.__protocol_attrs_cache__
+        except AttributeError:
+            protocol_attrs = _get_protocol_attrs(cls)
+            cls.__protocol_attrs_cache__ = protocol_attrs
+            return protocol_attrs
```

Comment:
We get this incorrect behaviour on your branch:

```pycon
>>> from typing import Protocol, runtime_checkable
>>> @runtime_checkable
... class Foo(Protocol):
...     x: int
...     
>>> @runtime_checkable
... class Bar(Foo, Protocol):
...     y: str
...     
>>> class HasX:
...     x = 42
...     
>>> isinstance(HasX(), Foo)
True
>>> isinstance(HasX(), Bar)
True
```

The result of the last `isinstance()` call there is incorrect, and is different from what we get on Python 3.13. The buggy behaviour occurs because the `cls.__protocol_attrs_cache__` lookup on `Bar` succeeds on the first access of `Bar.__protocol_attrs__` (it just finds the attribute on `Bar`'s supertype, `Foo` -- `Foo.__protocol_attrs_cache__` has already been populated due to the previous `isinstance(HasX(), Foo)` call). I think you need to do something like this instead:

```suggestion
        try:
            return cls.__dict__["__protocol_attrs_cache__"]
        except KeyError:
            protocol_attrs = _get_protocol_attrs(cls)
            cls.__protocol_attrs_cache__ = protocol_attrs
            return protocol_attrs
```

Reply:
Good catch, thanks. I used a slightly different solution to avoid using `__dict__`.